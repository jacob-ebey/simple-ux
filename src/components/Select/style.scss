@import '../variables';
@import 'mixin';

.select {
  display: flex;
  flex-direction: column;
  padding: $content-padding / 2 0;

  &-inline {
    flex-direction: row;
    align-items: center;

    &>label {
      margin-right: $content-padding / 2;
    }
  }

  &>select {
    height: $input-height + 6;
    min-width: 110px;

    @include select-box(
      $color-light,
      0,
      null,
      10px,
      $color-secondary,
      $color-dark,
      $color-light-darker,
      $color-dark,
      $color-dark,
      $border-width
    );
  }
}