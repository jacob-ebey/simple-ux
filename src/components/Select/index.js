import { h } from 'preact' // eslint-disable-line no-unused-vars

import { merge, mod } from '../tools'

import './style.scss'

let selectCount = 0

const Select = ({ label, items, inline, id, class: className, children, ...rest }) => {
  const selectId = id || `select_${selectCount++}`

  return (
    <div class={merge('select', mod('select', inline && 'inline'), className)}>
      { label && <label for={selectId}>{label}</label> }

      <select
        {...rest}
        id={selectId}
      >
        {children}
        {
          items && items.filter(item => typeof item !== 'undefined').map((item, index) => {
            let label = index
            let key = index

            if (typeof item !== 'object') {
              key = label = item
            } else {
              label = item.label || item.key
              key = item.key
            }

            return <option key={key} value={key}>{label}</option>
          })
        }
      </select>
    </div>
  )
}

export default Select
