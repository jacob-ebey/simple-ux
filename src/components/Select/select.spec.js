/* eslint-env jest */
import { h } from 'preact' // eslint-disable-line no-unused-vars
import { shallow } from 'preact-render-spy'

import Select from './'

const createItems = () => [
  { key: 1, label: 'Item 1' },
  { key: 2, label: 'Item 2' }
]

describe('Select', () => {
  describe('- snapshots -', () => {
    it('can render', () => {
      const wrapper = shallow(<Select />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can add custom class', () => {
      const wrapper = shallow(<Select class='custom' />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can render children', () => {
      const wrapper = shallow(
        <Select>
          <option>Option 1</option>
          <option>Option 2</option>
        </Select>
      )
      expect(wrapper).toMatchSnapshot()
    })

    it('can render label', () => {
      const wrapper = shallow(<Select label='Test Label' />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can render inline', () => {
      const wrapper = shallow(<Select inline />)
      expect(wrapper).toMatchSnapshot()
    })

    it('uses provided id', () => {
      const wrapper = shallow(<Select id='testid' />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can render string items', () => {
      const wrapper = shallow(<Select items={['1', '2']} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can render items', () => {
      const wrapper = shallow(<Select items={createItems()} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can handle undefined item', () => {
      const wrapper = shallow(<Select items={[0, undefined]} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can handle object without label', () => {
      const wrapper = shallow(<Select items={[{ key: 1 }]} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can render items and children', () => {
      const wrapper = shallow(
        <Select id='testid' items={createItems()}>
          <option>Default</option>
        </Select>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })
})
