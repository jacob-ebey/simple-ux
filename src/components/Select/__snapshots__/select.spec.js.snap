// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select - snapshots - can add custom class 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select custom">
  <select id="select_1"></select>
</div>

`;

exports[`Select - snapshots - can handle object without label 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="select_8">
    <option value={1}>1</option>
  </select>
</div>

`;

exports[`Select - snapshots - can handle undefined item 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="select_7">
    <option value={0}>0</option>
  </select>
</div>

`;

exports[`Select - snapshots - can render 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="select_0"></select>
</div>

`;

exports[`Select - snapshots - can render children 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="select_2">
    <option>Option 1</option>
    <option>Option 2</option>
  </select>
</div>

`;

exports[`Select - snapshots - can render inline 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select select-inline">
  <select id="select_4"></select>
</div>

`;

exports[`Select - snapshots - can render items 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="select_6">
    <option value={1}>Item 1</option>
    <option value={2}>Item 2</option>
  </select>
</div>

`;

exports[`Select - snapshots - can render items and children 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="testid">
    <option>Default</option>
    <option value={1}>Item 1</option>
    <option value={2}>Item 2</option>
  </select>
</div>

`;

exports[`Select - snapshots - can render label 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <label for="select_3">Test Label</label>
  <select id="select_3"></select>
</div>

`;

exports[`Select - snapshots - can render string items 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="select_5">
    <option value="1">1</option>
    <option value="2">2</option>
  </select>
</div>

`;

exports[`Select - snapshots - uses provided id 1`] = `
preact-render-spy (1 nodes)
-------
<div class="select">
  <select id="testid"></select>
</div>

`;
